discovery-provider/src/challenges/challenge_event_bus.py

```diff
@@ -40,21 +43,54 @@ def get_manager(self, challenge_id: str) -> ChallengeManager:
         """Gets a manager for a given challenge_id"""
         return self._managers[challenge_id]
 
+    @contextmanager
+    def use_scoped_dispatch_queue(self):
+        """Makes the bus only dispatch the events once out of the new scope created with 'with'"""
+        try:
+            self._use_in_memory_queue = True
+            yield self._in_memory_queue
+        finally:
+            logger.info(
+                f"ChallengeEventBus: Flushing {len(self._in_memory_queue)} events from in-memory queue"
+            )
+            self._use_in_memory_queue = False
+            for event in self._in_memory_queue:
+                self.dispatch(
+                    event["event"],
+                    event["block_number"],
+                    event["user_id"],
+                    event["extra"] if event["extra"] else {},
+                )
+            self._in_memory_queue.clear()
```

Comment:
Just to be extra paranoid, I don't think we should clear the entire queue here, since it's possible we have concurrent adds to the queue after we finish our dispatching for loop, which would then get cleared without ever getting dispatched. Definitely an edge case, but probably better to count the entries prior to the while loop and only clear those

Reply:
Hmm that's interesting - is this still a problem if we're using a different bus for each task?

Reply:
Only an issue if we have task reentry on the same thread because we lose a lock now, but that's so much less likely I'm not very worried about this anymore

Reply:
https://docs.celeryproject.org/en/stable/userguide/tasks.html#instantiation

> A task is not instantiated for every request, but is registered in the task registry as a global instance.

I could probably make a list of lists, might make sense anyway - inside of `use_scoped_dispatch_queue` get the number of in memory queues for our index and set to a local variable, add another queue to the list of queues, and then after the yield, only flushing the queue with that index